File: programmers/[프로그래머스] 주식가격.java
package programmers;

import java.util.*;


class Solution {
    public int[] solution(int[] prices) {
        
        //주식가격이 떨어지지 않았다 에서 -> stack 을 생각하는 키워드!
        
        int[] ans = new int[prices.length];
        Stack<Integer> stack = new Stack();
        
        for (int i = 0; i < prices.length; i++){
            while (!stack.isEmpty() && prices[i] < prices[stack.peek()]) {
                ans[stack.peek()] = i - stack.peek(); //주식가격이 떨어짐
                stack.pop(); //답을 구했기 때문에 stack에서 제거
            }
            stack.push(i);
        }
        
        while(!stack.isEmpty()){//값을 구하지 못한 index == 끝까지 가격이 떨어지지 않은 주식
            ans[stack.peek()] = (prices.length-1) - stack.peek();
            stack.pop();
        }
        
        return ans;
    }

}


---------------------------------

2021.08.06 다시 풀어봄
    
    
import java.util.*;


class Solution {
    public int[] solution(int[] prices) {
        
        Stack<Price> stack = new Stack<>();
        int[] answer = new int[prices.length];
        
        for( int idx = 0; idx < prices.length; idx++){
            while(!stack.isEmpty()){
                Price now = stack.peek();
                if(now.price() <= prices[idx])
                    break;
                answer[now.index()] = ( idx - now.index());
                stack.pop();
            }
            stack.push(new Price(prices[idx],idx));  
        }
            
        while(!stack.isEmpty()){
            Price price = stack.pop();
            answer[price.index()] = ( prices.length -1 - price.index() );
        }
        
        return answer;
        
    }
    
    public static class Price{
        
        int price;
        int index;
        
        public Price(int price, int index){
            this.price = price;
            this.index = index;
        }
        
        public int price(){
            return this.price;
        }
        
        public int index(){
            return this.index;
        } 
        
        public String print(){
            return String.valueOf(this.price) + String.valueOf(this.index);
        }
    }

}

File: programmers/[프로그래머스] 주식가격.py

#1. brute force
def solution(prices):
    #몇초인지
    n = len(prices)
    ans = [0]*n
    for i in range(n):
        for j in range(i+1,n):#만약 0초면 1초부터 비교
            if prices[i] > prices[j]:#주식가격이 떨어지면
                ans[i] += j-i
                break
        else:
            ans[i] = n-i-1
    return ans



--------------------------------------------------

# 2022- 05 -05
# 이게 더 빠름

def solution(prices):
    stack = []
    answer = [ 0 for i in range(len(prices))]
    
    for time, price in enumerate(prices):
        
        if len(stack) > 0:    
            while stack:
                if stack[-1][1] > price:
                    (s_time, s_price) = stack.pop()
                    answer[s_time] = time - s_time
                else: #주식 안떨어졌으면 
                    break
        
        stack.append((time,price))

    
    #끝까지 안떨어진 값들 정리
    cur_time = len(prices) - 1
    while stack:
        (time,price) = stack.pop()
        answer[time] = cur_time - time
        
-----------------------------------------------------
# stack에 index를 넣어준다.
def solution(prices):
    answer = [0]*len(prices)
    stack = []
    #index를 stack에 넣는다.
    
    for i, price in enumerate(prices):
        #stack에 있는 값이 현재값보다 크면 (값이 떨어지면)
        while stack and price < prices[stack[-1]]:
            j = stack.pop()
            answer[j] = i-j
        stack.append(i)
        
    while stack:
            j = stack.pop()
            answer[j] = len(prices) -1 -j
            
    return answer
